RecoVertex/PrimaryVertexProducer/interface/WeightedMeanFitter.h

```diff
@@ -0,0 +1,469 @@
+
+    GlobalPoint p(ox + t * vx, oy + t * vy, oz + t * vz);
+    return std::pair<GlobalPoint, double>(
+        p,
+        std::sqrt(std::pow(p.x() - vertex.x(), 2) + std::pow(p.y() - vertex.y(), 2) + std::pow(p.z() - vertex.z(), 2)));
+  }
+
+  inline TransientVertex weightedMeanOutlierRejection(const std::vector<std::pair<GlobalPoint, GlobalPoint>>& points,
+                                                      std::vector<reco::TransientTrack> iclus) {
+    float x = 0., y = 0., z = 0.;
+    float s_wx = 0., s_wz = 0.;
+    float s2_wx = 0., s2_wz = 0.;
+    float wx = 0., wz = 0., chi2 = 0.;
+    float ndof_x = 0.;
+
+    AlgebraicSymMatrix33 err;
+    err(0, 0) = startError / 10 * startError / 10;
+    err(1, 1) = startError / 10 * startError / 10;
+    err(2, 2) = startError * startError;  // error is 20 cm, so cov -> is 20 ^ 2
+    for (const auto& p : points) {
+      wx = p.second.x() <= precision ? 1. / std::pow(precision, 2) : 1. / std::pow(p.second.x(), 2);
+
+      wz = p.second.z() <= precision ? 1. / std::pow(precision, 2) : 1. / std::pow(p.second.z(), 2);
+
+      x += p.first.x() * wx;
+      y += p.first.y() * wx;
+      z += p.first.z() * wz;
+
+      s_wx += wx;
+      s_wz += wz;
+    }
+
+    if (s_wx == 0. || s_wz == 0.) {
+      edm::LogWarning("WeightedMeanFitter") << "Vertex fitting failed at beginning \n";
+      return TransientVertex(GlobalPoint(0, 0, 0), err, iclus, 0, 0);
+    }
+
+    x /= s_wx;
+    y /= s_wx;
+    z /= s_wz;
+
+    float old_x, old_y, old_z;
+
+    float xpull;
+    int niter = 0;
+    float mu = 3.;
+
+    float err_x, err_z;
+
+    err_x = 1. / s_wx;
+    err_z = 1. / s_wz;
+
+    while ((niter++) < 2) {
+      old_x = x;
+      old_y = y;
+      old_z = z;
+      s_wx = 0;
+      s_wz = 0;
+      s2_wx = 0;
+      s2_wz = 0;
+      x = 0;
+      y = 0;
+      z = 0;
+      ndof_x = 0;
+
+      for (unsigned int i = 0; i < (unsigned int)points.size(); i++) {
+        std::pair<GlobalPoint, double> p = nearestPoint(GlobalPoint(old_x, old_y, old_z), (iclus)[i].track());
+
+        wx = points[i].second.x() <= precision ? std::pow(precision, 2) : std::pow(points[i].second.x(), 2);
+        wz = points[i].second.z() <= precision ? std::pow(precision, 2) : std::pow(points[i].second.z(), 2);
+
+        xpull = 0.;
+
+        if (std::pow(p.first.x() - old_x, 2) / (wx + err_x) < mu * mu &&
+            std::pow(p.first.y() - old_y, 2) / (wx + err_x) < mu * mu &&
+            std::pow(p.first.z() - old_z, 2) / (wz + err_z) < mu * mu)
+          xpull = 1.;
+
+        ndof_x += xpull;
+
+        wx = xpull / wx;
+        wz = xpull / wz;
+
+        x += wx * p.first.x();
+        y += wx * p.first.y();
+        z += wz * p.first.z();
+
+        s_wx += wx;
+        s_wz += wz;
+
+        s2_wx += wx * xpull;
+        s2_wz += wz * xpull;
+      }
+
+      if (s_wx == 0. || s_wz == 0.) {
+        edm::LogWarning("WeightedMeanFitter") << "Vertex fitting failed" << s_wx << " , " << s_wz << "\n";
```

Comment:
this message can be improved.
Currently I see printouts as:
```
%MSG-w WeightedMeanFitter:   PrimaryVertexProducer:offlinePrimaryVerticesFromRefittedTrks  09-Nov-2022 16:04:36 CET Run: 1 Event: 2781
Vertex fitting failed0 , 0

%MSG
```

which looks difficult to parse.